### PR TITLE
Restore TemplateHaskell pragma in hls-graph

### PR DESCRIPTION
**Pull request by pepeiborra**
The pragma is required for the `embed-files` flag  and was removed recently in #2523.

@jneira @Anton-Latukha the PR is open for any ideas that extend CI to check the build with this flag.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2549"></a>

### CODE REVIEW

**Comment by Anton-Latukha** [comment]
I obviously have limited knowledge of the project.

Since I'm already sitting - would contribute to the patch.

The patch may uncover a bit of work.

**Comment by Anton-Latukha** [comment]
Oh. I thought we in CI do not test pre `ghcide` sub-project & so though to add a test-suites into the process.

But:
```
🅸 ~/s/h/c/haskell-language-server:fix-hls-graph-build⨯>cabal test shake-bench
No tests to run for the package shake-bench-0.1.0.2
🅸 ~/s/h/c/haskell-language-server:fix-hls-graph-build⨯>cabal test hls-graph
No tests to run for the package hls-graph-1.5.1.1
🅸 ~/s/h/c/haskell-language-server:fix-hls-graph-build⨯>cabal test hie-compat
No tests to run for the package hie-compat-0.2.1.0
🅸 ~/s/h/c/haskell-language-server:fix-hls-graph-build⨯>cabal test hls-plugin-api
No tests to run for the package hls-plugin-api-1.2.0.2
🅸 ~/s/h/c/haskell-language-server:fix-hls-graph-build⨯>cabal test hls-test-utils
No tests to run for the package hls-test-utils-1.1.0.2
🅸 ~/s/h/c/haskell-language-server:fix-hls-graph-build⨯>
```

There are no test-suites, no benchmarks for the.
So CI there is not of any help.

**Comment by pepeiborra** [comment]
Sorry I wasn't clear - we want to test that hls-graph builds with the  embed-files flag

**Comment by jneira** [comment]
well the unique thing I can think off without too much changes is add a step in the testing workflow building the package with the non default value for the flag
the package is not big and will not take too much time I hope

**Comment by michaelpj** [comment]
Possibly dumb question: how bad would it be to just use `file-embed` always?

**Comment by Anton-Latukha** [comment]
[skip circleci]

**Comment by Anton-Latukha** [comment]
`michaelpj` raised a sound question.

**Comment by Anton-Latukha** [comment]
The request/PR shows that there is a broad structurally important question. There is a number of subprojects & they have a number of flags. & CI does not tests any of them.

1. We can not test all flag combinations.
2. We can not test flag combinations * platform * GHC matrix.
3. That is why decided to create a workflow: 1) which tests for "default" case, under default currently took Linux & recommended GHC version 2) which builds one combination where all important flags are maximally enabled & so have some cross-testing.

**Comment by pepeiborra** [comment]
> Possibly dumb question: how bad would it be to just use `file-embed` always?

It introduces a dependency on TH, which makes HLS-graph less portable. Which probably doesn't matter much.

File: .github/workflows/caching.yml [hunk 1]
```diff
@@ -151,7 +151,7 @@ jobs:
           echo '' && \
           cat 'cabal.project.freeze' && \
           echo '' || \
-          echo 'WARNING: Could not produce the `freeze`.
```
Comment (Anton-Latukha) [inline .github/workflows/caching.yml:154]: whoopsie daisy

File: .github/workflows/caching.yml [hunk 2]
```diff
@@ -151,7 +151,7 @@ jobs:
...
+          echo 'WARNING: Could not produce the `freeze`.'
```
Comment (Anton-Latukha) [inline .github/workflows/caching.yml:154]: Whoopsie daisy

**Comment by Anton-Latukha** [comment]
The details (early termination rules) we can tweak in other PRs.

I consider this ready to be reviewed & merged.

**Comment by Anton-Latukha** [comment]
I would start CI configuration consolidation effort at https://github.com/haskell/haskell-language-server/issues/2422#issuecomment-1002279380.

File: .github/workflows/flags.yml [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+          cache-name: compiled-deps
+        with:
+          path: ${{ steps.HaskEnvSetup.outputs.cabal-store }}
+          key:  ${{ env.cache-name }}-${{ runner.os }}-${{ matrix.ghc }}-${{ env.INDEX_STATE }}-${{ hashFiles('cabal.project.freeze') }}
+          restore-keys: |
+                ${{ env.cache-name }}-${{ runner.os }}-${{ matrix.ghc }}-${{ env.INDEX_STATE }}-
+                ${{ env.cache-name }}-${{ runner.os }}-${{ matrix.ghc }}-
+                ${{ env.cache-name }}-${{ runner.os }}-
+
+      - if: needs.pre_job.outputs.should_skip != 'true'
```
Comment (jneira) [inline .github/workflows/flags.yml:134]: please add a `cabal update` here too, with the same comment as the other workflows if possible

File: .github/workflows/flags.yml [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        name: Build `hls-plugin-api` with flags
+        run: cabal v2-build hls-plugin-api --flags="pedantic"
+
+      - if: needs.pre_job.outputs.should_skip != 'true'
+        name: Build `hls-test-utils` with flags
+        run: cabal v2-build hls-test-utils --flags="pedantic"
+
+      # repeating builds to workaround segfaults in windows and ghc-8.8.4
+      - if: needs.pre_job.outputs.should_skip_ghcide != 'true'
```
Comment (jneira) [inline .github/workflows/flags.yml:151]: we are only in ubuntu here, so not need for three builds

File: .github/workflows/flags.yml [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                         , "install/**"
+                         , "**.nix"
+                         , "flake.lock"
+                         , "**/README.md"
+                         , "FUNDING.yml"
+                         , ".circleci/**"
+                         , "**/stack*.yaml"
+                         ]'
+      # If we only change ghcide downstream packages we have not test ghcide itself
+      - id: skip_ghcide_check
```
Comment (jneira) [inline .github/workflows/flags.yml:40]: afaics we are not using this condition in the workflow

File: .github/workflows/flags.yml [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        with:
+          path: ${{ steps.HaskEnvSetup.outputs.cabal-store }}
+          key:  ${{ env.cache-name }}-${{ runner.os }}-${{ matrix.ghc }}-${{ env.INDEX_STATE }}-${{ hashFiles('cabal.project.freeze') }}
+          restore-keys: |
+                ${{ env.cache-name }}-${{ runner.os }}-${{ matrix.ghc }}-${{ env.INDEX_STATE }}-
+                ${{ env.cache-name }}-${{ runner.os }}-${{ matrix.ghc }}-
+                ${{ env.cache-name }}-${{ runner.os }}-
+
+      - if: needs.pre_job.outputs.should_skip != 'true'
+        name: Build `hls-graph` with flags
```
Comment (jneira) [inline .github/workflows/flags.yml:135]:
```suggestion
      - name: Build `hls-graph` with flags
```

the job itself is skipped if the condition holds

File: .github/workflows/flags.yml [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        run: cabal v2-build hls-graph --flags="pedantic embed-files stm-stats"
+
+      - if: needs.pre_job.outputs.should_skip != 'true'
+        name: Build `hie-compat` with flags
```
Comment (jneira) [inline .github/workflows/flags.yml:139]:
```suggestion
      - name: Build `hie-compat` with flags
```

File: .github/workflows/flags.yml [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        run: cabal v2-build hie-compat --flags="ghc-lib"
+
+      - if: needs.pre_job.outputs.should_skip != 'true'
+        name: Build `hls-plugin-api` with flags
```
Comment (jneira) [inline .github/workflows/flags.yml:143]:
```suggestion
      - name: Build `hls-plugin-api` with flags
```

File: .github/workflows/flags.yml [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        run: cabal v2-build hls-plugin-api --flags="pedantic"
+
+      - if: needs.pre_job.outputs.should_skip != 'true'
+        name: Build `hls-test-utils` with flags
```
Comment (jneira) [inline .github/workflows/flags.yml:147]:
```suggestion
      - name: Build `hls-test-utils` with flags
```

File: .github/workflows/flags.yml [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        run: cabal v2-build hls-test-utils --flags="pedantic"
+
+      # repeating builds to workaround segfaults in windows and ghc-8.8.4
+      - if: needs.pre_job.outputs.should_skip_ghcide != 'true'
+        name: Build
```
Comment (jneira) [inline .github/workflows/flags.yml:152]:
```suggestion
      - name: Build
```

**Review by jneira — CHANGES_REQUESTED** [review]

File: .github/workflows/flags.yml (hunk 4 above)
Comment (Anton-Latukha) [reply .github/workflows/flags.yml:151]: Yes.

I still do this with deduplication in mind (doing deduplication locally).

**Comment by Anton-Latukha** [comment]
This state confuses me.

Seems like all requested changes were addressed & the PR is still in the `change requested` mode.

I rechecked a couple of times & can not tell what is the cause.

**Comment by jneira** [comment]
if a revision requests changes the revision author (in this case me) has to approve manually the pr to make it mergeable

**Review by jneira — APPROVED** [review]
To have a new workflow is nice for correctness and completeness but we have too much workflow filed so I would try to deduplicate workflows before adding more

**Comment by Anton-Latukha** [comment]
The difficulty with computation would start where there would no be new build workers for PRs, so far it looks like CI have more then enough build workers, so far I haven't seen waits on resources.

We had `caching` workflow added, which runs async and has short-circuiting. `caching` saves computational time overall.

HLint workflow runs one worker & takes ~1m to run.

Flags runs one worker ~5m.

Testing takes 15 workers, takes 3-4 of them ~90m & Benchmark takes 3 workers, 2 workers for ~95m.

So we only have a code complexity & shotgun surgery because of CI design (but most CIs have alike yaml-based design). And looking at DHall results (but I would still work on it a bit before reporting) - the current shotgun surgery situation in CI starts to look quite good.

**Pull request merged**